### PR TITLE
Mesh 554/claim rules multiple matching

### PR DESCRIPTION
**Pull request by fmiguelgarcia**
# Multiple matching.
It supports multiple matching rules over a Claim, e.g. `Jurisdiction`. In the new architecture, the **evaluation of claims is externalized** (kind of visitor pattern) to a new couple of structures: `Predicate` & `predicate::Context`.  They provide the _flexibility_ to support _any new requirement_ about claims evaluation.

# Claim fetching optimization
The new fetching layer allows to **iterate over claims and reduce the memory foot-print** on specific cases. It also allows fetching claims faster because the _storage structure is aligned with the use case of claim evaluation_. That case is extremely critical because is executed in almost every transaction.
The new way to store claims is using a `double_map` where the first key is `(target_DID, ClaimType)` and the second one is ( `Issuer_DID`, `Optional<Scope>`). In that way, getting CDD for a specific ID is straightforward, and it allows to iterate over all associated CDD claims.

# New Unit Test: Jurisdiction asset transmission rule.
See `fn jurisdiction_asset_rules_we` in `general_tm.rs`

 # More stuff
- README.md has been updated with "debug instructions".
- Remove some `Self::is_signer_authorized`: That functionality is covered by `Context::current_identity_or` function.
- Some clippy/warning suggestions have been fixed.

### CODE REVIEW

File: primitives/src/predicate.rs [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+use sp_std::prelude::*;
+
+/// Context using during an `Predicate` evaluation.
+///
+/// # TODO
+///  - Use a lazy access to claims. It could be part of the optimization
+///  process of CDD claims.
+#[derive(Encode, Decode, Clone, Debug, Default)]
+pub struct Context {
```
Comment (vkomenda) [inline primitives/src/predicate.rs]: "Context" is not really a lucky choice of word here. Predicates map values of their variables to true or false. Quantifiers map predicates to true or false. Those two work at a different level. Usually by context people mean something different than either values or predicates. It's actually more like assignment of values to free variables and a list of axioms. I'm not going to nag on this but this is the thing.

In addition, "and", "or" and "not" map propositions (things that are either true or false) to propositions, rather than values to true/false. But that's OK in your context because by `and` you mean "apply these two predicates to the same subject". A bit confusing though.

I suggest renaming `Context` as `Value` or `Subject`. Also references to quantifiers like `any` or `exists` better be removed because there aren't any quantifiers really, just predicates again.

File: primitives/src/predicate.rs [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+/// It creates a predicate to evaluate the existential of `claim` in the context.
+#[inline]
+pub fn exists(claim_type: ClaimType) -> ExistentialPredicate {
+    ExistentialPredicate { claim_type }
+}
+
+/// It creates a predicate to evaluate if any of `claims` are found in the context.
+#[inline]
+pub fn any(claims: Vec<Claim>) -> AnyPredicate {
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply primitives/src/predicate.rs]: I used any because it is quite similar to the concept of `iter::any`. 
I wouldn't use `Intersects` because it could implies to return the intersection instead of a boolean.

File: primitives/src/predicate.rs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    pub claim_type: ClaimType,
+}
+
+impl Predicate for ExistentialPredicate {
+    #[inline]
+    fn evaluate(&self, context: &Context) -> bool {
+        context
+            .claims
+            .iter()
+            .any(|ctx_claim| ctx_claim.claim_type() == self.claim_type)
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply primitives/src/predicate.rs]: This is because the close relation between `Claim` and `ClaimType`.  The current status only support one claim with data: Jurisdiction.

File: primitives/src/predicate.rs [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        self.lhs.evaluate(context) || self.rhs.evaluate(context)
+    }
+}
+
+// NotPredicate
+// ======================================================
+
+/// Predicate that returns a logical NOT of other predicate.
+#[derive(Encode, Decode, Clone, Debug)]
+pub struct NotPredicate<P: Predicate + Sized> {
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply primitives/src/predicate.rs]: It's a side-effect because 'AndPredicate' & 'OrPredicate' are built from functions 'and' & 'or', whose have the `Size` constraint.
I am going to add this type constraint.

File: primitives/src/identity_claim.rs [hunk 5]
```diff
@@ -1,38 +1,81 @@
...
-pub enum IdentityClaimData {
+pub enum Claim {
     /// User is Accredited
-    Accredited(IdentityId),
+    Accredited,
     /// User is Accredited
-    Affiliate(IdentityId),
+    Affiliate,
     /// User has an active BuyLockup (end date defined in claim expiry)
-    BuyLockup(IdentityId),
```
Comment (maxsam4) [inline primitives/src/identity_claim.rs]: Why was the target did removed from all claims?

File: primitives/src/rule.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+/// It defines the type of rule supported, and the filter information we will use to evaluate as a
+/// predicate.
+pub enum RuleType {
+    /// Rule to ensure that claim filter produces one claim.
+    IsPresent(ClaimType),
+    /// Rule to ensure that claim filter produces an empty list.
+    IsAbsent(ClaimType),
+    /// Rule to ensure that at least one claim is fetched when filter is applied.
+    IsAnyOf(Vec<Claim>),
+}
```
Comment (maxsam4) [inline primitives/src/rule.rs]: I think we also need `IsNoneOf(Vec<Claim>)`.

File: primitives/src/predicate.rs (hunk 1 above)
Comment (fmiguelgarcia) [reply primitives/src/predicate.rs]: The idea behind "Context" is that "rule" defines the filters to apply on fetching claims. That is why I discarded other names like 'Value'. 
Right now it stores values in memory, but there is an optimization task on Jira, in which we could explore the possibility of use directly the iterator extracted from fetching phase.
As a context, we could also add other information that future predicates would need.

File: primitives/src/identity_claim.rs (hunk 5 above)
Comment (fmiguelgarcia) [reply primitives/src/identity_claim.rs]: A merge mistake! Thanks @maxsam4 .

Comment (fmiguelgarcia) [reply primitives/src/identity_claim.rs]: Because it was duplicated info:

    Target identifier is already stored as part of the first key in Identity::Claims (double_map).
    Issuer identifier is already stored as part of second key in Identity::Claims (double_map).

Using that new Identity::Claims definition, we improve the fetching of claims.

Comment (maxsam4) [reply primitives/src/identity_claim.rs]: I think there's a confusion here, we need three dids and calling all of them target is turning out to be a bad idea xD. The 3 dids required represent:
1) Claim Receiver (part of the first key in Identity::Claims)
2) Claim Issuer (part of the second key in Identity::Claims)
3) Target/Context/Relevant DID. This will be either the did of a token or of a token issuer. This was stored in claims in form of `BuyLockup(IdentityId)`.

The idea is that "Alice" will be able to issuer a claim on "Bob" saying "Bob" is Whitelisted in "Charlier's token".
"Alice" is claim issuer
"Bob" is claim receiver
"Charlier's token" is Target/Context/Relevant DID.

File: primitives/src/rule.rs (hunk 6 above)
Comment (fmiguelgarcia) [reply primitives/src/rule.rs]: `IsNoneOf` has been added.
The predicate implementation is the following: 
```
RuleType::IsNoneOf(claims) => not(any(claims)).evaluate(context),
```

File: primitives/src/predicate.rs (hunk 1 above)
Comment (vkomenda) [reply primitives/src/predicate.rs]: In that case it's better not to call functions taking context to true/false predicates. Would calling them `Rule`s be possible? That would be better for things like joining predicates ("and") and alternating predicated ("or") as that would remove the unwanted connotations with predicate logic which this rule system is not.

File: pallets/identity/src/lib.rs [hunk 7]
```diff
@@ -95,8 +122,8 @@ decl_storage! {
         /// It stores the current identity for current transaction.
         pub CurrentDid: Option<IdentityId>;
 
-        /// (DID, claim_data, claim_issuer) -> Associated claims
-        pub Claims: double_map hasher(blake2_256) IdentityId, blake2_256(ClaimIdentifier) => IdentityClaim;
+        /// (Target ID, claim type) (issuer,scope) -> Associated claims
+        pub Claims: double_map hasher(blake2_256) Claim1stKey, blake2_256(Claim2ndKey) => IdentityClaim;
```
Comment (maxsam4) [inline pallets/identity/src/lib.rs]: It will be best if the first key only contains the target id. This is because the dApp needs to show all claims that an identity has. If the first key is just the target identity, we can easily enumerate all its claims.

Alternatively, we need to add an RPC function that loops over all claim types to fetch all claims of an identity and then return the combined result.

**Comment by maxsam4** [comment]
I like this PR overall but the whole predicate system feels over-engineered for our use case. That being said, it is neat and we might end up using it in other things as well in the future.

File: primitives/src/identity_claim.rs (hunk 5 above)
Comment (fmiguelgarcia) [reply primitives/src/identity_claim.rs]: @maxsam4 I used the following semantic for claims:
1.  Claim Receiver -> **target**
2.  Claim Issuer -> **issuer**
3.  Target/Context/Relevant DID -> **scope**

File: pallets/identity/src/lib.rs (hunk 7 above)
Comment (fmiguelgarcia) [reply pallets/identity/src/lib.rs]: I vote for option 2: Iterate over claims, because this is the performance key to optimize the following operations:

-  Fetching CDD, which is going to be a very frequency operation. In this case, first valid CDD will be enough.
-  Evaluating Claims in GTM (or any other TM) is based on rules which are defined per claim type.

Comment (maxsam4) [reply pallets/identity/src/lib.rs]: @fmiguelgarcia If I remember correctly, We don't iterate over all CDD or any other claim type in any situation on-chain. We always fetch claims from trusted providers by generating both keys and that takes O(n) time. If we move `claim_type` to `Claim2ndKey`, we'll still be able to fetch claims in O(n) time.

File: primitives/src/rule.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+/// Type of claim requirements that a rule can have
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct Rule {
+    /// Type of rule.
+    pub rule_type: RuleType,
+    /// Trusted issuers.
+    pub issuers: Vec<IdentityId>,
+    /// Claim scope
+    pub scope: Option<IdentityId>,
```
Comment (maxsam4) [inline primitives/src/rule.rs]: I think scope should be part of the `rule_type` so that people can do something like "IsAnyOf whitelist with scope A or whitelist with scope B".

File: pallets/identity/src/lib.rs (hunk 7 above)
Comment (fmiguelgarcia) [reply pallets/identity/src/lib.rs]: @maxsam4 Actually, it is just the opposite. Let see common use-cases:
-  Fetch CDD Claims. It this case we search by `Target Identity` & `ClaimType::CDD`, and we iterate over result. e.g. in `identity/src/lib.rs` functions like `has_valid_cdd` or `fetch_cdd`.
   - `has_valid_cdd` stops at first claim which was issued by a valid CDD. There is a high probability of being the first element, since only CDD providers are allowed to create CDD claims (another task).
   - `fetch_cdd` It is similar to `has_valid_cdd` but it returns the CDD provider Id. It also has a `leeway` parameter, that we apply to Claim expiration. That's the main reason why this code is not similar to `has_valid_cdd` but their behaviours are mostly the same.
- Rules evaluation in GTM.  Each rule uses a specific `ClaimType`, so we are in the same situation, we have to fetch claims for a `Target Identity` + `ClaimType`. Optionally we can filter more using the claim `scope` and/or the `issuer` of the claim.  That optimization is in `general_tm.rs`, function `fetch_context`.

In summary, both cases use the same fist key (`Claim1stKey`), and the rule evaluation process can use even more filtering (some cases will use `Claim2ndKey`), reducing the fetching size from disk and the memory footprint.

Comment (maxsam4) [reply pallets/identity/src/lib.rs]: In none of the cases, we iterate over all results of only Claim1stKey. We always fetch claims from trusted issuer and scope. The scope is not needed in most cases but trusted issuer is needed and it will become a dos vector if we were to fetch all claims of `Claim1stKey` without trusted issuer because anyone will be able to issue a hundred claims on someone and then we'll have to iterate over all those claims to find our trusted claim. This is why we always iterate over trusted issuers and fetch single claims issued by them. The same concept applies to GTM.

I can't find any location where we are iterating solely over `Claim1stKey`. Can you please provide a specific line of code where we are doing this?

Comment (maxsam4) [reply pallets/identity/src/lib.rs]: Seems like the logic was changed in this PR. I must've missed it during the review. Due to the attack vector I mentioned above, we can't do this. We can not iterate over just Claim1stKey because it can become an unbounded array.

File: primitives/src/rule.rs (hunk 8 above)
Comment (fmiguelgarcia) [reply primitives/src/rule.rs]: That is a pretty nice use case. However if we add `scope` into `rule_type`, the evaluation of non-scoped claims (like CDD) would be more complex (at least on my mind).
My suggestion is to extend Rule to support AND, OR operations (similar to Predicate) which will be more expressive and flexible.
So for instance:
```
 Rule { 
    rule_type: IsAnyOf ( x, y, z),
    scope: 'A' 
}. or( Rule {
    rule_type: IsAnyOf( x', y', z'),
    scope: 'B'
}). and( Rule {
    rule_type: IsPresent(CDD)
}
```

BTW in order to limit this task, we can evaluate that extensions/ideas on another task.

File: pallets/identity/src/lib.rs [hunk 9]
```diff
@@ -1232,106 +1234,125 @@ impl<T: Trait> Module<T> {
...
+    /// It iterates over all claims of type `claim_type` for target `id` identity.
+    /// Please note that it could return expired claims.
+    fn fetch_base_claims<'a>(
+        target: IdentityId,
+        claim_type: ClaimType,
+        scope: &'a Option<IdentityId>,
+    ) -> impl Iterator<Item = IdentityClaim> + 'a {
+        let pk = Claim1stKey { target, claim_type };
+
+        <Claims>::iter_prefix(pk).filter(move |c| Self::is_identity_claim_in_scope(c, scope))
```
Comment (maxsam4) [inline pallets/identity/src/lib.rs]: We should not be iterating over claims based on just the Claim1stKey because anyone can issue claims and increase the number of claim under a specific Claim1stKey. This becomes a dos vector where Alice can issue thousands of CDD claim on Bob but since Alice is not a CDD provider, all those claims will be invalid and Bob might have to iterate over all those fake claims to finally get to his valid cdd claim.

File: primitives/src/rule.rs (hunk 8 above)
Comment (maxsam4) [reply primitives/src/rule.rs]: I think this is one of the main product requirements and needs to be implemented. That being said, I agree that it might make sense to do in a different task. cc @adamdossa WDYT?

File: pallets/identity/src/lib.rs (hunk 9 above)
Comment (fmiguelgarcia) [reply pallets/identity/src/lib.rs]: Only CDD providers are going to be able to produce CDD Claims and adding a claim is going to have a cost, so attacker is going to spend a lot of money, and Bob could just remove all its claims of an specific issuer (attacker) on a cheap and single transaction (TBD).

BTW It sounds quite interesting to limit the case of a rich attacker, e.g:
- The cost of adding a claim based on the size of the number of elements contained in `Claim1stKey`. Let say constant for regular case (< 20 claims), linear incremental for busy target identity ( >= 20 & <50 claims), exponential for attackers( >50 claims).

Again this case seems out of the box of this task and we could manage it on an specific security one.

Comment (maxsam4) [reply pallets/identity/src/lib.rs]: Claim receivers can not remove claims. Only claim issuers can remove claims so bob will never be able to remove Alice's claim.

This "attack" is being introduced by this PR so I think it makes sense to solve within this PR. That being said, I understand that this PR has already been stretched enough and I'll be fine if we create a new task to tackle this problem but it's something we need to tackle before V1 release. cc @adamdossa

File: pallets/identity/src/lib.rs (hunk 7 above)
Comment (fmiguelgarcia) [reply pallets/identity/src/lib.rs]: @maxsam4 Unbounded claim fetch has been removed. We have **only 2 public entry points**:
-  `fetch_claim( target: IdentityId, claim_type: ClaimType, issuer: IdentityId, scope: Option<Scope> )`. It returns a `Option<IdentityClaim>`.
- `fetch_cdd( target: IdentityId, leeway: u64) -> Option<IdentityId>`. It fetches the first valid CDD claim.

**Scope** is an alias of `IdentityId`. It is an option to support claim indexation without scope (like CDD), but it is **not** a wildcard.
